### PR TITLE
[ot] scripts/opentitan: add a GDB replay script from QEMU exec log

### PR DESCRIPTION
**Pull request by rivos-eblot**

### CODE REVIEW

File: docs/opentitan/gdbreplay.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                        Address to load each specified binary
+  -b BIN, --bin BIN     Binary application
+  -g GDB, --gdb GDB     GDB server (default to localhost:3333)
+  -v, --verbose         increase verbosity
+  -d, --debug           enable debug mode
+````
+
+### Arguments
+
+* `-a` specify an address where to load the matching RAW binary application, see `-b` optiion for
```
Comment (loiclefort) [inline docs/opentitan/gdbreplay.md]:
```suggestion
* `-a` specify an address where to load the matching RAW binary application, see `-b` option for
```

File: scripts/opentitan/gdbreplay.py [hunk 2]
```diff
@@ -0,0 +1,899 @@
...
+
+           :param efp: a File-like (binary read access)
+        """
+        # use a copy of the stream to release the file pointer.
+        try:
+            self._elf = ELFFile(BytesIO(efp.read()))
+        except ELFError as exc:
+            raise ValueError(f'Invalid ELF file: {exc}') from exc
+        if self._elf['e_machine'] != 'EM_RISCV':
+            raise ValueError('Not an RISC-V ELF file')
```
Comment (loiclefort) [inline scripts/opentitan/gdbreplay.py]:
```suggestion
            raise ValueError('Not a RISC-V ELF file')
```

File: scripts/opentitan/gdbreplay.py [hunk 3]
```diff
@@ -0,0 +1,899 @@
...
+    def entry_point(self) -> Optional[int]:
+        """Provide the entry point of the application, if any.
+
+           :return: the entry point address
+        """
+        return self._elf and self._elf.header.get('e_entry', None)
+
+    @property
+    def raw_size(self) -> int:
+        """Provide the size of the Secure Boot Header section, if any.
```
Comment (loiclefort) [inline scripts/opentitan/gdbreplay.py]:
```suggestion
        """Provide the size of the payload, if any.
```

File: scripts/opentitan/gdbreplay.py [hunk 4]
```diff
@@ -0,0 +1,899 @@
...
+            else:
+                if phy_addr > phy_start+size:
+                    self._log.debug('fill gap with previous segment')
+                    size = phy_addr-phy_start
+            size += seg_size
+        if phy_start is None:
+            raise ValueError('No loadable segment found')
+        return phy_start, size
+
+    def _build_payload(self) -> bytes:
```
Comment (loiclefort) [inline scripts/opentitan/gdbreplay.py:187]: `_parse_segments` and `_build_payload` look very similar, maybe factorize common code?

File: scripts/opentitan/gdbreplay.py [hunk 5]
```diff
@@ -0,0 +1,899 @@
...
+                raise RuntimeError('Reached end of exec stream')
+        else:
+            if self._xpos > 0:
+                self._xpos -= 1
+            else:
+                raise RuntimeError('Reached start of exec stream')
+
+    def cont(self, back: bool = False, addr: Optional[int] = None) -> bool:
+        """Continue execution of instruction stream till either a HW breakpoint
+           or the end of the execution streamed is reached.
```
Comment (loiclefort) [inline scripts/opentitan/gdbreplay.py]:
```suggestion
           or the end of the execution stream is reached.
```

File: scripts/opentitan/gdbreplay.py [hunk 6]
```diff
@@ -0,0 +1,899 @@
...
+        for lno, line in enumerate(qfp, start=1):
+            tmo = self.TCRE.match(line)
+            if not tmo:
+                continue
+            scpu, spc, func = tmo.groups()
+            xcpu = int(scpu)
+            xpc = int(spc, 16)
+            if not lno % 10000:
+                self._log.debug('Parsed %d lines', lno)
+            #self._log.debug("0x%08x %s", xpc, func)
```
Comment (loiclefort) [inline scripts/opentitan/gdbreplay.py]: leftover debug trace?

File: scripts/opentitan/gdbreplay.py [hunk 7]
```diff
@@ -0,0 +1,899 @@
...
+            end = buf.find(b'#', start)
+            if end < 0:
+                continue
+            if len(buf)-end < 2:
+                continue
+            req = buf[start+1:end]
+            crc = int(buf[end+1:end+3], 16)
+            buf = req[end+3:]
+            # if start != 0:
+            #     continue
```
Comment (loiclefort) [inline scripts/opentitan/gdbreplay.py]: dead code?

File: scripts/opentitan/gdbreplay.py [hunk 8]
```diff
@@ -0,0 +1,899 @@
...
+    def _do_query_c(self, *_) -> str:
+        return f'QC{self._thread_id}'
+
+    def _do_query_supported(self, payload: bytes) -> str:
+        """Query supported features."""
+        req = payload.decode()
+        resp = [f'PacketSize={self.MAX_PACKET_LENGTH-16:x}',
+                'ReverseStep+', 'ReverseContinue+']
+        for cap in req.split(';'):
+            # supp = cap in ('hwbreak+', 'vContSupported+', 'no-resumed+')
```
Comment (loiclefort) [inline scripts/opentitan/gdbreplay.py]: leftover tests?

Comment (rivos-eblot) [reply scripts/opentitan/gdbreplay.py]: fixed

File: scripts/opentitan/gdbreplay.py (hunk 7 above)
Comment (rivos-eblot) [reply scripts/opentitan/gdbreplay.py]: fixed

File: scripts/opentitan/gdbreplay.py (hunk 6 above)
Comment (rivos-eblot) [reply scripts/opentitan/gdbreplay.py]: fixed

File: scripts/opentitan/gdbreplay.py (hunk 5 above)
Comment (rivos-eblot) [reply scripts/opentitan/gdbreplay.py]: fixed

File: scripts/opentitan/gdbreplay.py (hunk 3 above)
Comment (rivos-eblot) [reply scripts/opentitan/gdbreplay.py]: fixed

File: scripts/opentitan/gdbreplay.py (hunk 2 above)
Comment (rivos-eblot) [reply scripts/opentitan/gdbreplay.py]: fixed

File: docs/opentitan/gdbreplay.md (hunk 1 above)
Comment (rivos-eblot) [reply docs/opentitan/gdbreplay.md]: fixed

File: scripts/opentitan/gdbreplay.py (hunk 4 above)
Comment (rivos-eblot) [reply scripts/opentitan/gdbreplay.py:187]: We do not want to build the full object in RAM when we only need to compute the final size.

**Review by loiclefort — APPROVED** [review]
LGTM

**Pull request merged**